server.go

```diff
@@ -1587,14 +1589,28 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
 // utilize the ordering of the local and remote public key. If we didn't use
 // such a tie breaker, then we risk _both_ connections erroneously being
 // dropped.
-func shouldDropLocalConnection(local, remote *btcec.PublicKey) bool {
+func (s *server) shouldDropLocalConnection(local, remote *btcec.PublicKey) bool {
 	localPubBytes := local.SerializeCompressed()
 	remotePubPbytes := remote.SerializeCompressed()
 
 	// The connection that comes from the node with a "smaller" pubkey
 	// should be kept. Therefore, if our pubkey is "greater" than theirs, we
 	// should drop our established connection.
-	return bytes.Compare(localPubBytes, remotePubPbytes) > 0
+	supposedToDrop := bytes.Compare(localPubBytes, remotePubPbytes) > 0
+
+	pubStr := string(remotePubPbytes)
+
+	// Lookup the current sequence number for this peer. If no entry is
+	// found, we start in state 0. We take the states mod 3 so that we
+	// the sequence eventually terminates, without making assumptions about
+	// the starting states of the peers.
+	state := s.peerSequence[pubStr]
+	s.peerSequence[pubStr] = (state + 1) % 3
+
+	// Stay honest to the protocol only if we are in state 0.
+	stayHonest := state == 0
+
+	return supposedToDrop == stayHonest
```

Comment:
This is very hard to understand without having read the description of the protocol beforehand. Maybe better to write code/comments as  "we choose `A<B` in state 0, and `B>A` in state 1 and 2"?

Reply:
If you squint, this is actually computing the xor of the two values. Though it could be documented/explained better :)